src/main/kotlin/com/zoltu/gradle/plugin/GitVersioning.kt

```diff
@@ -64,25 +72,59 @@ class GitVersioning : Plugin<Project> {
 	}
 
 	private fun setProjectVersion(project: Project, versionInfo: VersionInfo) {
-		project.version = versionInfo.toString()
+		project.version = configuration.customVersionToString.apply(versionInfo)
 	}
 
 	private fun setJarManifestVersion(project: Project, versionInfo: VersionInfo) {
 		val closure = object {
 			@Suppress("unused")
 			fun apply(jar: Jar) {
 				val attributes = jar.manifest?.attributes ?: throw Exception("The Jar task has no manifest.")
-				attributes.put("Implementation-Version", versionInfo.toString())
+				attributes.put("Implementation-Version", configuration.customVersionToString.apply(versionInfo))
 				attributes.put("Implementation-Sha", versionInfo.sha)
 				attributes.put("Specification-Version", "${versionInfo.major}.${versionInfo.minor}")
 			}
 		}
 		(project.tasks ?: throw Exception("The project has no tasks.")).withType(Jar::class.java, MethodClosure(closure, "apply"))
 	}
 
-	private fun exposeVersionInfoToUser(project: Project, versionInfo: VersionInfo) {
-		project.extensions.create("ZoltuGitVersioning", Extension::class.java, versionInfo)
-	}
+	private fun exposeConfigurationObject(project: Project, configurationChangeCallback: () -> Unit) = project.extensions.create("ZoltuGitVersioning", Configuration::class.java, configurationChangeCallback)
 
-	open class Extension(open val versionInfo: VersionInfo)
+	open class Configuration(private val configurationChangeCallback: () -> Unit) {
+		companion object {
+			@Suppress("unused")
+			const val NAME = "versioning"
+			private val defaultVersionToString = Function<VersionInfo, String> { versionInfo: VersionInfo ->
+				val suffix = if (versionInfo.tags != null && versionInfo.patch != null) {
+					// semantic versioning
+					"${versionInfo.patch}-${versionInfo.tags}.${versionInfo.commitCount}"
+				} else if (versionInfo.patch != null) {
+					// semantic versioning
+					"${versionInfo.patch}-${versionInfo.commitCount}"
+				} else if (versionInfo.tags != null) {
+					// simple versioning
+					"${versionInfo.commitCount}-${versionInfo.tags}"
+				} else {
+					// simple versioning
+					"${versionInfo.commitCount}"
+				}
+				"${versionInfo.major}.${versionInfo.minor}.$suffix".toString()
+			}
+			private val defaultCustomDescribeProcessor = Function<String, VersionInfo?> { null }
+		}
+		/**
+		 * Contains version information for consumption.
+		 */
+		lateinit var versionInfo: VersionInfo
+
+		/**
+		 * A custom describe processor that will take in a string as input and return a VersionInfo object as output.
+		 */
+		var customDescribeProcessor: Function<String, VersionInfo?> by observable(defaultCustomDescribeProcessor, { property, old, new -> configurationChangeCallback() })
```

Comment:
A little more research lead me to find that named captures are not available in Kotlin until 1.1, due to the current tight coupling with JDK 6 which doesn't support named captures: https://youtrack.jetbrains.com/issue/KT-12753

I attempted to switch the describe processor from returning `VersionInfo` to returning a map only to realize that the way things are setup now is to fail fast, so if the describe output doesn't initially start with something that matches one of the existing patterns then the plugin will fail before the user gets a chance to apply configuration.  The correct way to resolve this is to not apply versioning until *after* the gradle file has been parsed.  However, if I make that change then the `VersionInfo` data won't be available until after the gradle file has been parsed, which means you wouldn't be able to do `version "${ZoltuGitVersioning.versionInfo.major}..." in your `build.gradle` file.

I could switch it to silently fail, but that leads to user confusion when their project isn't setup right.  I could continue on with not evaluating the versioning until after `build.gradle` has been evaluated and require users to set whatever they want in a closure, but that leads to `build.gradle` complexity for simple scenarios.

I'm going to think on this some more over-night.  At the least, we can get the change you wanted in, which would allow users to specify a custom `toString()` method.  Unlike your workaround, this would make it so the version printed into the JAR manifest would also match your desired structure, and `./gradlew version` would print the desired version.